html/modules/custom/hr_paragraphs/hr_paragraphs.module

```diff
@@ -2497,6 +2497,85 @@ function hr_paragraphs_node_presave(EntityInterface $entity) {
   }
 }
 
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function hr_paragraphs_paragraph_presave(EntityInterface $entity) {
+  /** @var \Drupal\paragraphs\Entity\Paragraph $entity */
+  if (!$entity->hasField('field_text')) {
+    return;
+  }
+
+  // Only if format is basic_html.
+  if (!$entity->field_text->format) {
+    return;
+  }
+
+  if ($entity->field_text->format != 'basic_html') {
+    return;
+  }
+
+  $text = $entity->field_text->value;
+  if (empty($text)) {
+    return;
+  }
+
+  // Use DOMDocument to reliably clean HTML.
+  $dom = new \DOMDocument();
+
+  // Suppress warnings for malformed HTML.
+  libxml_use_internal_errors(TRUE);
+
+  // Load HTML fragment.
+  $dom->loadHTML('<?xml encoding="utf-8" ?>' . $text, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
+  libxml_clear_errors();
+
+  // Remove unwanted attributes from all elements.
+  foreach ($dom->getElementsByTagName('*') as $el) {
+    // Remove class and style attributes.
+    $el->removeAttribute('class');
+    $el->removeAttribute('style');
+
+    // Remove data-* attributes and empty attributes.
+    $attrs = [];
+    foreach ($el->attributes as $attr) {
+      $attrs[] = $attr->name;
+    }
+    foreach ($attrs as $attr_name) {
+      if (strpos($attr_name, 'data-') === 0 || $el->getAttribute($attr_name) === '') {
+        $el->removeAttribute($attr_name);
+      }
+    }
+  }
+  // Remove empty tags.
+  // We need to iterate multiple times in case of nested empty tags.
+  $removed = TRUE;
+  while ($removed) {
+    $removed = FALSE;
+    foreach ($dom->getElementsByTagName('*') as $el) {
+      if (!$el->hasChildNodes() && trim($el->textContent) === '') {
+        $el->parentNode->removeChild($el);
+        $removed = TRUE;
+        break;
```

Comment:
Is this intentional? It means sibling empty tags at the same level for example would not be removed in the foreach. That would results in more calls to `getElementsByTagName` and iterations to effectively remove those tags.